xds/src/main/java/io/grpc/xds/InterLocalityPicker.java

```diff
@@ -29,53 +29,62 @@
 
 final class InterLocalityPicker extends SubchannelPicker {
 
-  private final List<WeightedChildPicker> weightedChildPickers;
+  private final List<IntraLocalityPicker> intraLocalityPickers;
   private final ThreadSafeRandom random;
   private final int totalWeight;
 
-  static final class WeightedChildPicker {
-    final int weight;
-    final SubchannelPicker childPicker;
+  static final class IntraLocalityPicker extends SubchannelPicker {
```

Comment:
We want to verify the composition of the spawned `InterLocalityPicker` at each picker update, so we want `WeightedChildPicker` to carry locality information. Since that, `WeightedChildPicker` now is really an `IntraLocalityPicker` instead of a wrapper that hides this information.

Reply:
You can test the behavior at pickSubchannel. `....getWeight()).isEqualTo(...)` in your current test is not the right way to test.

Reply:
The inter-locality picking logic is faked out in `LocalityStoreTest`, and it's separately tested in `InterLocalityPickerTest`. I think checking the weight directly is the better way of testing. Maybe it's better to separate the refactor and the fix in two different PRs.

Reply:
We can still check weight indirectly with faked out inter-locality picking. We don't need use any real inter locality picker.

Reply:
@dapengzhang0 was saying to not add locality information into `WeightedChildPicker`, instead to proceed till `Subchannel`s are picked out and use picked `Subchannel`s to infer corresponding localities. With that, we have the mapping between locality and its corresponding `WeightChildPicker`.

We would need to replace all usages of `FakePickerFactory`'s `nextIndex` with this approach as relying on the knowledge of how `List<WeightChildPickers>` is ordered for localities is not correct and makes the test extremely fragile.

Again, in my opinion, this introduces a lot of complexity for testing the weight account for each locality. This makes `LocalityStore`'s test even messier. I am proposing for relating locality information with each in-locality picker (that is, make `WeightedChildPicker` `IntraLocalityPicker` and we are definitely able to get the locality an `IntraLocalityPicker` belonging to) in implementation.

BTW, `WeightChildPicker` not extending `SubchannelPicker` looks bad. It looks wired to do `WeightChildPicker#childPicker#pickSubchannel(...)`. But I failed to convince the author. :weary: